### PR TITLE
todo enumerated

### PR DESCRIPTION
**Pull request by strorinWind**
Workshop 3 about clickListener

### CODE REVIEW

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                    parent.context
+                ).inflate(R.layout.item_actors_data, parent, false)
+            )
+        }
+    }
+
+    override fun onBindViewHolder(holder: ActorsViewHolder, position: Int) {
+        when (holder) {
+            is DataViewHolder -> {
+                holder.onBind(imageOption, actors[position])
```
Comment (emartynov) [inline app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt]: I would have default empty function bind() in parent and replace this when with one line

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ActorsViewHolder {
+        return when (viewType) {
+            VIEW_TYPE_EMPTY -> EmptyViewHolder(
+                LayoutInflater.from(parent.context)
+                    .inflate(R.layout.item_actors_empty, parent, false)
+            )
+            else -> DataViewHolder(
+                LayoutInflater.from(
```
Comment (emartynov) [inline app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt:33]: Minor - prefer dot wrapping and parameters when having a fluent interface

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        actors = newActors.toMutableList()
+    }
+}
+
+abstract class ActorsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
+
+private class EmptyViewHolder(itemView: View) : ActorsViewHolder(itemView)
+private class DataViewHolder(itemView: View) : ActorsViewHolder(itemView) {
+
+    val avatar: ImageView? = itemView.findViewById(R.id.iv_actor_avatar)
```
Comment (emartynov) [inline app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt]: private?

Comment (emartynov) [reply app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt]: why optional? Aren't we sure these views on item view?

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import com.android.fundamentals.R
+import com.android.fundamentals.data.models.Actor
+import com.android.fundamentals.workshop03.solution.ActorsViewHolder
+import com.bumptech.glide.Glide
+import com.bumptech.glide.request.RequestOptions
+
+//TODO 2: make listener constructor parameter
+class WS03ActorsAdapter: RecyclerView.Adapter<ActorsViewHolder>() {
+
+    private val imageOption = RequestOptions()
```
Comment (emartynov) [inline app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt]: Why are image options are part of the adapter and not view holder?

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsFragment.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    override fun onDetach() {
+        recycler = null
+
+        super.onDetach()
+    }
+
+    private fun updateData() {
+        (recycler?.adapter as? WS03ActorsAdapter)?.apply {
+            bindActors(ActorsDataSource().getActors())
+            notifyDataSetChanged()
```
Comment (emartynov) [inline app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsFragment.kt]: Why notify is outside of the adapter? Isn't it functionality related to adapter?

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsFragment.kt [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    private fun doOnClick(actor: Actor) {
+        recycler?.let { rv ->
+            Snackbar
+                .make(rv, "You chose ${actor.name}", Snackbar.LENGTH_SHORT)
```
Comment (emartynov) [inline app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsFragment.kt]: String reesources?

**Review by emartynov — COMMENTED** [review]
Some minor comments

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt (hunk 2 above)
Comment (strorinWind) [reply app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt:33]: Not really important

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt (hunk 1 above)
Comment (strorinWind) [reply app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt]: Not sure that we can fully get rid of the casting if we do not want to set clickListener to EmptyViewHolder

Comment (emartynov) [reply app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt]: Understand! It depends on the case but if there are multiple actions that could come from the user then I implement it in way that view holder exposes stream of possible events and adapter or another entity subscribes to it. That makes a flexible reactive contract and some view holders will just expose stream that never produces anything. But also the lifecycle of this stream is subject to take care when the holder is reattached from the screen.

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt [hunk 7]
```diff
@@ -65,19 +59,24 @@ abstract class ActorsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemVi
 private class EmptyViewHolder(itemView: View) : ActorsViewHolder(itemView)
 private class DataViewHolder(itemView: View) : ActorsViewHolder(itemView) {
 
-    val avatar: ImageView? = itemView.findViewById(R.id.iv_actor_avatar)
-    val name: TextView? = itemView.findViewById(R.id.tv_actor_name)
-    val oscarState: TextView? = itemView.findViewById(R.id.tv_actor_oscar_state)
+    private val imageOption = RequestOptions()
```
Comment (emartynov) [inline app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsAdapter.kt]: A small optimisation - you can move it to companion

File: app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsFragment.kt [hunk 8]
```diff
@@ -53,8 +53,10 @@ class WS03ActorsFragmentSolution : Fragment() {
 
     private fun doOnClick(actor: Actor) {
         recycler?.let { rv ->
-            Snackbar
-                .make(rv, "You chose ${actor.name}", Snackbar.LENGTH_SHORT)
+            Snackbar.make(
+                        rv,
+                        rv.context.getString(R.string.fragment_actors_chosen_text, actor.name),
```
Comment (emartynov) [inline app/src/main/java/com/android/fundamentals/workshop03/WS03ActorsFragment.kt]: The fragment has its own `getString()` method you don't need any context

**Pull request merged**